src/UI/Actions.hs

```diff
@@ -438,52 +437,46 @@ selectedItemHelper
     :: Applicative f
     => Getting (L.List n t) AppState (L.List n t)
     -> AppState
-    -> (t -> f (AppState -> AppState))
+    -> ((Int, t) -> f (AppState -> AppState))
     -> f AppState
 selectedItemHelper l s func =
   ($ s) <$> case L.listSelectedElement (view l s) of
-  Just (_, m) -> func m
+  Just m -> func m
   Nothing -> pure $ setError (GenericError "No item selected.")
 
-applyItemTags
-    :: (ListItemSetter a, Notmuch.ManageTags a, MonadIO f)
-    => a
-    -> [Text]
-    -> ([Text] -> a -> a)
-    -> AppState
-    -> f (AppState -> AppState)
-applyItemTags m ts op s =
+getEditorTagOps :: Lens' AppState (E.Editor Text Name) -> AppState -> [TagOp]
+getEditorTagOps widget s =
+  let contents = (unlines $ E.getEditContents $ view widget s)
+  in foldOf folded (parseTagOps contents)
+
+applyTagOps
+  :: (MonadIO m)
+  => [TagOp]
+  -> Vector.Vector (Int, NotmuchMail)
+  -> AppState
+  -> m (Either Error (Vector.Vector (Int, NotmuchMail)))
+applyTagOps ops mails s =
   let dbpath = view (asConfig . confNotmuch . nmDatabase) s
-  in either setError updateListItem <$> runExceptT (Notmuch.writeTags dbpath (op ts m))
-
-applyEditorMailTags :: (ListItemSetter a, Notmuch.ManageTags a) => a -> AppState -> IO (AppState -> AppState)
-applyEditorMailTags m s =
-  let widget = if view asAppMode s == ManageMailTags then (asMailIndex . miMailTagsEditor) else (asMailIndex . miThreadTagsEditor)
-      contents = (unlines $ E.getEditContents $ view widget s)
-      tags = strip <$> splitOn "," contents
-  in applyItemTags m tags Notmuch.setTags s
+  in runExceptT (Notmuch.messageTagModify dbpath ops mails)
 
 updateStateWithParsedMail :: AppState -> IO AppState
-updateStateWithParsedMail s = selectedItemHelper (asMailIndex . miListOfMails) s $ \m ->
+updateStateWithParsedMail s = selectedItemHelper (asMailIndex . miListOfMails) s $ \(_, m) ->
         either
             (\e -> setError e . set asAppMode BrowseMail)
             (\pmail -> set (asMailView . mvMail) (Just pmail) . set asAppMode ViewMail)
             <$> runExceptT (parseMail m (view (asConfig . confNotmuch . nmDatabase) s))
 
-updateReadState :: ([Text] -> NotmuchMail -> NotmuchMail) -> AppState -> IO AppState
+updateReadState :: TagOp -> AppState -> IO AppState
 updateReadState op s =
-  let newTag = view (asConfig . confNotmuch . nmNewTag) s
-  in selectedItemHelper (asMailIndex . miListOfMails) s $ \m -> applyItemTags m [newTag] op s
+  selectedItemHelper (asMailIndex . miListOfMails) s $ \m ->
+  either setError updateMails <$> applyTagOps [op] (Vector.singleton m) s
+
+updateMails :: Vector.Vector (Int, NotmuchMail) -> AppState -> AppState
+updateMails mails = over (asMailIndex . miListOfMails . L.listElementsL) (flip Vector.update mails)
```

Comment:
[nit] HLint suggests `Use section`:
```haskell
(`Vector.update` mails)
```
I tend to agree, whether it's `Vector.update` or `safeUpdate` :)

Reply:
Done. (Also, fixed my missing hlint invocation)